### ISSUE TITLE
Quick Question Regarding Image Captioning Loss Function

### ISSUE THREAD

**Issue by abaybektursun** [OFA-Sys/OFA#204]
I am tying to train image captioning model, however my captions are not full sentences but rather list of short descriptions (1-4 words) of the items in the image. I have a lot of data.
So for example `img1.jpg` -> `"curved gray monitor", "old style lamp", "black mouse"`. Note that you can shuffle the descriptions in that list, and they still would be the same meaning.
My questions is, **what loss function should I use**? I was thinking regular cross entropy but it will try to exactly match the tokens, in my case however it does not matter whether output is `"curved gray monitor", "old style lamp", "black mouse"` or `"black mouse", "old style lamp", "curved gray monitor"`.

**Comment by abaybektursun** [comment]
In your `pretrain_data_examples.zip` you have a file `detection_examples.tsv`. 
I looked at one of the labels:
```
['505.856',
 '189.994',
 '799.744',
 '450.016',
 '/m/07j7r',
 'tree&&753.664',
 '260.596',
 '1012.736',
 '301.35',
 '/m/0c9ph5',
 'flower&&22.528',
......]
```
This does not seem to be visual grounding since there is no text prompt. Do you have an example script for training on this (`detection_examples.tsv`) type of dataset? What loss you use?

**Comment by abaybektursun** [comment]
Looks like this object detection data is only used for pre-training using [adjust_label_smoothed_cross_entropy](https://github.com/OFA-Sys/OFA/blob/2480865f3cc005e3cb12974d33bd78f79f0b661b/run_scripts/pretraining/pretrain_ofa_base.sh#L29).
After some searching I fund out that [DETR](https://arxiv.org/pdf/2005.12872.pdf) uses exactly the loss I was looking for: [bipartite matching loss](https://parthpankajtiwary.medium.com/end-to-end-object-detection-with-transformers-direct-set-prediction-8da7dddfd18d). I wonder if I can implement this loss and extend it as a fairseq criterion?

**Comment by logicwong** [comment]
@abaybektursun For detection, we following [pix2seq](https://arxiv.org/abs/2109.10852) to quantinize the original coordinates into discretized values, so we can use simple cross-entropy for this task. Back to your question, I think you can try the following two approaches:

1. Divide the origin sample into different image-text pairs, such as (img1.jpg, "curved gray monitor"), (img1.jpg, "old style lamp"), (img1.jpg, "black mouse"), and then use cross-entropy loss function for training. After training, you can use [Diverse Beam Search](https://github.com/OFA-Sys/OFA/blob/2480865f3cc005e3cb12974d33bd78f79f0b661b/models/search.py#L551-L618) provided by Fairseq to generate different results for each image.

2. Just treat the original sample as a single training sample, but at each epoch you can shuffle the objects in each image like (epoch1, img1.jpg, "curved gray monitor ; old style lamp ; black mouse"), (epoch2, img1.jpg, "black mouse", "old style lamp", "curved gray monitor"), and then also train your model using the cross-entropy loss function. After training, you can adjust the generation args to control the results, such as [generation length](https://github.com/OFA-Sys/OFA/blob/2480865f3cc005e3cb12974d33bd78f79f0b661b/fairseq/fairseq/dataclass/configs.py#L769-L783), [length penalty](https://github.com/OFA-Sys/OFA/blob/2480865f3cc005e3cb12974d33bd78f79f0b661b/fairseq/fairseq/dataclass/configs.py#L796-L801), etc. This method is similar to pix2seq, I think it can help you to generate multiple objects in one result.

If you use [adjust_label_smoothed_cross_entropy](https://github.com/OFA-Sys/OFA/blob/2480865f3cc005e3cb12974d33bd78f79f0b661b/run_scripts/pretraining/pretrain_ofa_base.sh#L29), remember setting drop-worst-ratio to 0.0, I think this arg will hurt the performance in this dataset. In addition, there are some other tricks that can help you improve performance, such as [R-Drop](https://github.com/OFA-Sys/OFA/blob/2480865f3cc005e3cb12974d33bd78f79f0b661b/criterions/label_smoothed_cross_entropy.py#L46-L48) and [CIDEr optimization](https://github.com/OFA-Sys/OFA/blob/main/criterions/scst_loss.py), you can try.

As for [bipartite matching loss](https://parthpankajtiwary.medium.com/end-to-end-object-detection-with-transformers-direct-set-prediction-8da7dddfd18d), it is a non-autoregressive solution that has been successfully applied to object detection tasks, but I'm not sure if it works for image captioning. You can try the above two methods first. If they don't work well, you can use bipartite matching loss, and I'd love to hear the results.

**Comment by abaybektursun** [comment]
Thanks! I will let you know how it goes.
